Maple2.Server.Game/Model/Field/Actor/ActorState/AnimationState.cs

```diff
@@ -0,0 +1,220 @@
+﻿using Maple2.Model.Metadata;
+using Maple2.Server.Core.Packets;
+using Maple2.Server.Game.Model.Enum;
+
+namespace Maple2.Server.Game.Model.Field.Actor.ActorState;
+
+public class AnimationState {
+    // TODO: maybe make this a ping based variable?
+    private const long ClientGraceTimeTick = 500; // max time to allow client to go past loop & sequence end
+    private readonly IActor actor;
+
+    private struct LoopData {
+        public float start;
+        public float end;
+
+        public LoopData(float start, float end) {
+            this.start = start;
+            this.end = end;
+        }
+    }
+
+    public struct TickPair {
+        public long server;
+        public long client;
+
+        // mobs
+        public TickPair(long server) {
+            this.server = server;
+            this.client = server;
+        }
+
+        // players
+        public TickPair(long server, long client) {
+            this.server = server;
+            this.client = client;
+        }
+    }
+
+    public AnimationMetadata? RigMetadata { get; init; }
+    public AnimationSequence? PlayingSequence { get; private set; }
+    private bool IsPlayerAnimation { get; set; }
+    public float MoveSpeed { get; set; }
+    public float AttackSpeed { get; set; }
+    private float sequenceSpeed { get; set; }
+    private float lastSequenceTime { get; set; }
+    private float sequenceEnd { get; set; }
+    private LoopData sequenceLoop { get; set; }
+    private long sequenceEndTick { get; set; }
+    private long sequenceLoopEndTick { get; set; }
+    private long lastTick { get; set; }
+    private bool isLooping { get; set; }
+    private bool loopOnlyOnce { get; set; }
+    private AnimationType sequenceType { get; set; }
+
+    private bool debugPrintAnimations;
+    public bool DebugPrintAnimations {
+        get { return debugPrintAnimations; }
+        set {
+            if (actor is FieldPlayer) {
+                debugPrintAnimations = value;
+            }
+        }
+    }
+
+    public AnimationState(IActor actor, string modelName) {
+        this.actor = actor;
+
+        RigMetadata = actor.NpcMetadata.GetAnimation(modelName);
+        MoveSpeed = 1;
+        AttackSpeed = 1;
+        IsPlayerAnimation = actor is FieldPlayer;
+    }
+
+    private void ResetSequence() {
+        PlayingSequence = null;
+        sequenceSpeed = 1;
+        lastSequenceTime = 0;
+        sequenceLoop = new LoopData(0, 0);
+        lastTick =0;
```

Comment:
nit, to this never happen again. you should format on save.
```suggestion
        lastTick = 0;
```